File: lib/iam_stack.py
"""iam_stack.py
This module defines the IAMStack class, which creates IAM roles and policies
for the CI/CD pipeline components including CodePipeline and CodeBuild.
"""

from typing import Optional

import aws_cdk as cdk
from aws_cdk import aws_iam as iam
from constructs import Construct


class IAMStackProps(cdk.StackProps):
  """
  IAMStackProps defines the properties for the IAMStack.
  
  Args:
    environment_suffix (Optional[str]): Environment suffix for resource naming
    **kwargs: Additional keyword arguments passed to the base cdk.StackProps
  """
  
  def __init__(self, environment_suffix: Optional[str] = None, **kwargs):
    super().__init__(**kwargs)
    self.environment_suffix = environment_suffix


class IAMStack(cdk.NestedStack):
  """
  IAMStack creates IAM roles and policies for CI/CD pipeline components.
  
  This stack creates:
  - CodePipeline service role
  - CodeBuild service role  
  - CloudFormation execution role for deployments
  
  Args:
    scope (Construct): The parent construct
    construct_id (str): The unique identifier for this stack
    props (Optional[IAMStackProps]): Optional properties for configuring the stack
    **kwargs: Additional keyword arguments passed to the CDK NestedStack
  """
  
  def __init__(
    self,
    scope: Construct,
    construct_id: str,
    props: Optional[IAMStackProps] = None,
    **kwargs
  ):
    super().__init__(scope, construct_id, **kwargs)
    
    environment_suffix = props.environment_suffix if props else 'dev'
    
    # CodePipeline service role
    self.codepipeline_role = iam.Role(
      self,
      "CodePipelineRole",
      role_name=f"ciapp-{environment_suffix}-codepipeline-role",
      assumed_by=iam.ServicePrincipal("codepipeline.amazonaws.com"),
      managed_policies=[
        iam.ManagedPolicy.from_aws_managed_policy_name("AWSCodePipelineFullAccess"),
        iam.ManagedPolicy.from_aws_managed_policy_name("AWSCodeBuildDeveloperAccess"),
        iam.ManagedPolicy.from_aws_managed_policy_name("AmazonS3FullAccess"),
        iam.ManagedPolicy.from_aws_managed_policy_name("AWSCloudFormationFullAccess")
      ],
      inline_policies={
        "PassRolePolicy": iam.PolicyDocument(
          statements=[
            iam.PolicyStatement(
              effect=iam.Effect.ALLOW,
              actions=["iam:PassRole"],
              resources=["*"],
              conditions={
                "StringEquals": {
                  "iam:PassedToService": [
                    "cloudformation.amazonaws.com",
                    "codebuild.amazonaws.com"
                  ]
                }
              }
            )
          ]
        )
      }
    )
    
    # CodeBuild service role
    self.codebuild_role = iam.Role(
      self,
      "CodeBuildRole", 
      role_name=f"ciapp-{environment_suffix}-codebuild-role",
      assumed_by=iam.ServicePrincipal("codebuild.amazonaws.com"),
      managed_policies=[
        iam.ManagedPolicy.from_aws_managed_policy_name("CloudWatchLogsFullAccess"),
        iam.ManagedPolicy.from_aws_managed_policy_name("AmazonS3FullAccess")
      ],
      inline_policies={
        "CodeBuildPolicy": iam.PolicyDocument(
          statements=[
            iam.PolicyStatement(
              effect=iam.Effect.ALLOW,
              actions=[
                "logs:CreateLogGroup",
                "logs:CreateLogStream", 
                "logs:PutLogEvents"
              ],
              resources=[
                f"arn:aws:logs:us-west-2:{cdk.Aws.ACCOUNT_ID}:"
                f"log-group:/aws/codebuild/*"
              ]
            )
          ]
        )
      }
    )
    
    # CloudFormation execution role for deployments
    self.cloudformation_role = iam.Role(
      self,
      "CloudFormationRole",
      role_name=f"ciapp-{environment_suffix}-cloudformation-role", 
      assumed_by=iam.ServicePrincipal("cloudformation.amazonaws.com"),
      managed_policies=[
        iam.ManagedPolicy.from_aws_managed_policy_name("PowerUserAccess")
      ],
      inline_policies={
        "IAMPolicy": iam.PolicyDocument(
          statements=[
            iam.PolicyStatement(
              effect=iam.Effect.ALLOW,
              actions=[
                "iam:*"
              ],
              resources=["*"]
            )
          ]
        )
      }
    )
    
    # Add tags to all resources
    cdk.Tags.of(self).add("Environment", environment_suffix)
    cdk.Tags.of(self).add("Component", "IAM")

File: lib/iam_stack.go
package lib

import (
	"github.com/aws/aws-cdk-go/awscdk/v2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsiam"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type IamStackProps struct {
	*awscdk.StackProps
	EnvironmentName string
}

type IamStack struct {
	awscdk.Stack
	Ec2Role         awsiam.Role
	InstanceProfile awsiam.InstanceProfile
}

func NewIamStack(scope constructs.Construct, id *string, props *IamStackProps) *IamStack {
	var sprops awscdk.StackProps
	if props.StackProps != nil {
		sprops = *props.StackProps
	}
	stack := awscdk.NewStack(scope, id, &sprops)

	// Create IAM role for EC2 instance following least privilege principle
	ec2Role := awsiam.NewRole(stack, jsii.String("EC2InstanceRole"), &awsiam.RoleProps{
		RoleName:    jsii.String("ProductionEC2InstanceRole"),
		AssumedBy:   awsiam.NewServicePrincipal(jsii.String("ec2.amazonaws.com"), nil),
		Description: jsii.String("IAM role for EC2 instance with minimal required permissions"),
		ManagedPolicies: &[]awsiam.IManagedPolicy{
			// Basic EC2 permissions for SSM Session Manager
			awsiam.ManagedPolicy_FromAwsManagedPolicyName(jsii.String("AmazonSSMManagedInstanceCore")),
		},
	})

	// Add custom inline policy for Secrets Manager access with least privilege
	secretsPolicy := awsiam.NewPolicyDocument(&awsiam.PolicyDocumentProps{
		Statements: &[]awsiam.PolicyStatement{
			awsiam.NewPolicyStatement(&awsiam.PolicyStatementProps{
				Effect: awsiam.Effect_ALLOW,
				Actions: &[]*string{
					jsii.String("secretsmanager:GetSecretValue"),
					jsii.String("secretsmanager:DescribeSecret"),
				},
				Resources: &[]*string{
					jsii.String("arn:aws:secretsmanager:us-east-1:*:secret:prod/database/credentials-*"),
					jsii.String("arn:aws:secretsmanager:us-east-1:*:secret:prod/app/config-*"),
				},
			}),
			awsiam.NewPolicyStatement(&awsiam.PolicyStatementProps{
				Effect: awsiam.Effect_ALLOW,
				Actions: &[]*string{
					jsii.String("kms:Decrypt"),
				},
				Resources: &[]*string{
					jsii.String("arn:aws:kms:us-east-1:*:key/*"),
				},
				Conditions: &map[string]interface{}{
					"StringEquals": map[string]*string{
						"kms:ViaService": jsii.String("secretsmanager.us-east-1.amazonaws.com"),
					},
				},
			}),
		},
	})

	awsiam.NewPolicy(stack, jsii.String("SecretsManagerAccessPolicy"), &awsiam.PolicyProps{
		PolicyName: jsii.String("ProductionSecretsManagerAccess"),
		Document:   secretsPolicy,
		Roles:      &[]awsiam.IRole{ec2Role},
	})

	// Create instance profile
	instanceProfile := awsiam.NewInstanceProfile(stack, jsii.String("EC2InstanceProfile"), &awsiam.InstanceProfileProps{
		InstanceProfileName: jsii.String("ProductionEC2InstanceProfile"),
		Role:                ec2Role,
	})

	// Tag IAM resources
	awscdk.Tags_Of(ec2Role).Add(jsii.String("Environment"), jsii.String("Production"), nil)
	awscdk.Tags_Of(ec2Role).Add(jsii.String("Component"), jsii.String("IAM"), nil)

	return &IamStack{
		Stack:           stack,
		Ec2Role:         ec2Role,
		InstanceProfile: instanceProfile,
	}
}
